### PR TITLE
[On dev branch] Tuple rewrite

### PR DESCRIPTION
**Pull request by lxuechen**
Contents in this PR:
- Remove support for tuple `y0` and tuple-based `sde` for `sdeint`.
  - Rationale: This feature is somewhat niche. Supporting this would likely require adding an alternative tuple-based `ForwardSDE` that has the correct versions of `f`, `g_prod`, `gdg_prod`, and `dg_ga_jvp` (per our email discussion). 
- Make `AdjointSDE` take in `ForwardSDE` and use the latter's `g_prod`.
  - This simplifies `g_prod` of the former.
- Make `AdjointSDE` take in (and output) flattened tensors and perform flattening/unflattening inside the registered functions. 
  - One advantage of this is that we only need to unflatten 2 tensors now. 
- Refactor solvers to not be based on tuples.
- Explicit flattening/unflattening in `adjoint.py`
  - This differs slightly from the current implementation in `torchdiffeq`, where the `_TupleFunc` construction is in `_check_inputs`.
  - Rationale: We assume that `sdeint.integrate` only supports tensor-based `y0` and `sde`, and we should enforce this usage for now. 
- Filter out the parameters that don't require gradients in `sdeint_adjoint`.
- Fix plot folder names for files in `diagnostics`.
- *Temporarily* deprecating `logqp` support. 
  - This should either be added back or the example latent_sde rewritten with state augmentation when we get the general noise adjoint working with the log-ODE scheme. (I view getting the log-ODE scheme-based adjoint working for general noise Strat SDE as a milestone. )
- Fix scalar noise shape check and it's functionality in `sdeint` inconsistency. Hence, resolving #27.

The rate diagnostics files run, but I haven't been able to obtain rates due to #36. I need to run these when this is fixed. All tests pass.

Overall, this PR should also make adding `dg_ga_jvp` for the adjoint easier.

### CODE REVIEW

File: tests/test_strat.py [hunk 1]
```diff
@@ -75,38 +75,35 @@ def _batch_jacobian(output, input_):
     return torch.stack(jacs, dim=0)
 
 
-def _gdg_jvp_brute_force(sde, t, y, a):
+def _dg_ga_jvp_brute_force(sde, t, y, a):
     # Only returns the value for the first input-output pair.
```
Comment (patrick-kidger) [inline tests/test_strat.py:79]: Is this referring to the re-rewrite potential of having tuples? If so it can be removed.

File: torchsde/_core/adjoint.py [hunk 2]
```diff
@@ -167,24 +70,16 @@ def backward(ctx, *grad_outputs):
         dt_min = ctx.dt_min
         adjoint_options = ctx.adjoint_options
 
-        params = misc.make_seq_requires_grad(sde.parameters())
-        n_tensors, n_params = len(ans), len(params)
-
+        aug_state = (ys[-1], grad_ys[-1], *[torch.zeros_like(param) for param in params])
```
Comment (patrick-kidger) [inline torchsde/_core/adjoint.py:73]: Nitpick: I think this would be slightly neater as `(ys[-1], grad_ys[-1]) + tuple(torch.zeros_like(param) for param in params)`.

File: torchsde/_core/adjoint.py [hunk 3]
```diff
@@ -195,27 +90,22 @@ def backward(ctx, *grad_outputs):
                 dt_min=dt_min,
                 options=adjoint_options
             )
+            aug_state = misc.flat_to_shape(aug_state[1], shapes)  # Unpack the state at time -ts[i - 1].
+            aug_state[0] = ys[i - 1]
+            aug_state[1] += grad_ys[i - 1]
```
Comment (patrick-kidger) [inline torchsde/_core/adjoint.py:95]: In principle is `sdeint_adjoint` going to support double backward? I've been thinking that this may actually be a very important thing for us to support wrt our current project.
(I hadn't mentioned it as I didn't want to add to your workload. But here everything looks like it should pretty much work. At least in principle; not sure how accurate the double-gradients will actually be.)

I'd suggest changing this to an out-of-place update just in case, and changing the call to `sdeint.integrate` to `_SdeintAdjointMethod.apply`, and I think we should be good to go.

File: torchsde/_core/base_sde.py [hunk 4]
```diff
@@ -48,117 +48,116 @@ def __init__(self, sde):
         if hasattr(sde, "h"):
```
Comment (patrick-kidger) [inline torchsde/_core/base_sde.py:48]: This can be removed now.

File: torchsde/_core/base_sde.py [hunk 5]
```diff
@@ -48,117 +48,116 @@ def __init__(self, sde):
...
+            dg_ga_jvp = dg_ga_jvp.diagonal(dim1=-2, dim2=-1).sum(-1)
+        return dg_ga_jvp
 
-    def h(self, t, y):
-        return self._base_sde.h(t, y[0]),
+    def _return_zero(self, t, y, v):  # noqa
+        return 0.
 
 
 class RenameMethodsSDE(BaseSDE):
```
Comment (patrick-kidger) [inline torchsde/_core/base_sde.py:163]: This doesn't need `h` any more.

File: torchsde/_core/methods/milstein.py [hunk 6]
```diff
@@ -50,55 +49,47 @@ def v_term(self, I_k, dt):
...
-        g_prod_eval = self.sde.g_prod(t0, y0, I_k)
+        f = self.sde.f(t0, y0)
+        g_prod_I_k = self.sde.g_prod(t0, y0, I_k)
 
         if self.options[METHOD_OPTIONS.grad_free]:
-            g_eval = self.sde.g(t0, y0)
-            g_prod_eval_v = self.sde.g_prod(t0, y0, v)
+            g = self.sde.g(t0, y0)
+            g_prod_v = self.sde.g_prod(t0, y0, v)
             sqrt_dt = torch.sqrt(dt) if isinstance(dt, torch.Tensor) else math.sqrt(dt)
```
Comment (patrick-kidger) [inline torchsde/_core/methods/milstein.py:66]: I think `dt` can only be a tensor now (as input times are now coerced to tensor), so this can be simplified.

File: torchsde/_core/methods/srk.py [hunk 7]
```diff
@@ -63,38 +59,31 @@ def diagonal_or_scalar_step(self, t0, t1, y0):
         dt = t1 - t0
         sqrt_dt = torch.sqrt(dt) if isinstance(dt, torch.Tensor) else math.sqrt(dt)
         I_k, I_k0 = self.bm(t0, t1, return_U=True)
-        I_kk = [(delta_bm_ ** 2. - dt) / 2. for delta_bm_ in I_k]
-        I_kkk = [(delta_bm_ ** 3. - 3. * dt * delta_bm_) / 6. for delta_bm_ in I_k]
+        I_kk = (I_k ** 2 - dt) / 2.
+        I_kkk = (I_k ** 3 - 3. * dt * I_k) / 6.
```
Comment (patrick-kidger) [inline torchsde/_core/methods/srk.py:63]: Can these divisions be substituted for multiplications? (Making `_r6 = 1/6` a global constant is a little cleaner; writing out `0.16666...67` to full precision is a little faster; IMO doesn't really matter which is done.)

File: torchsde/_core/methods/srk.py [hunk 8]
```diff
@@ -63,38 +59,31 @@ def diagonal_or_scalar_step(self, t0, t1, y0):
...
-                    H0s_ + srid2.A0[s][j] * f_eval_ * dt + srid2.B0[s][j] * g_eval_ * I_k0_ / dt
-                    for H0s_, f_eval_, g_eval_, I_k0_ in zip(H0s, f_eval, g_eval, I_k0)
-                ]
-                H1s = [
-                    H1s_ + srid2.A1[s][j] * f_eval_ * dt + srid2.B1[s][j] * g_eval_ * sqrt_dt
-                    for H1s_, f_eval_, g_eval_ in zip(H1s, f_eval, g_eval)
-                ]
+                f = self.sde.f(t0 + srid2.C0[j] * dt, H0[j])
+                g = self.sde.g(t0 + srid2.C1[j] * dt, H1[j])
+                g = g.squeeze(2) if g.dim() == 3 else g
```
Comment (patrick-kidger) [inline torchsde/_core/methods/srk.py:72]: This is going to break if/when we support multiple batch dimensions. (Plus it's a bit opaque what it's doing.) Factor this out to a different method for additive/diagonal a la Milstein?

File: torchsde/_core/misc.py [hunk 9]
```diff
@@ -109,12 +115,26 @@ def grad(outputs, inputs, **kwargs):
...
+
+def flat_to_shape(tensor, shapes, length=()):
+    tensor_list = []
+    total = 0
+    for shape in shapes:
+        next_total = total + shape.numel()
+        # It's important that this be view((...)), not view(...). Else when length=(), shape=() it fails.
+        tensor_list.append(tensor[..., total:next_total].view((*length, *shape)))
+        total = next_total
+    return tensor_list
```
Comment (patrick-kidger) [inline torchsde/_core/misc.py:140]: There's now several unused functions in this file, that can be removed.

File: torchsde/_core/sdeint.py [hunk 10]
```diff
@@ -17,44 +17,41 @@
...
-from .._brownian import BaseBrownian, TupleBrownian, BrownianInterval
+from . import misc
+from .._brownian import BaseBrownian, BrownianInterval
 from ..settings import SDE_TYPES, NOISE_TYPES, METHODS, LEVY_AREA_APPROXIMATIONS
 from ..types import TensorOrTensors, Scalar, Vector
 
 
-def sdeint(sde: [base_sde.BaseSDE],
+def sdeint(sde: base_sde.BaseSDE,
            y0: TensorOrTensors,
```
Comment (patrick-kidger) [inline torchsde/_core/sdeint.py:29]: This should just be a tensor now. I've not checked the whole codebase exhaustively, but do we need TensorOrTensors any more?

File: torchsde/_core/sdeint.py [hunk 11]
```diff
@@ -17,44 +17,41 @@
...
+           method: Optional[str] = "srk",
            dt: Optional[Scalar] = 1e-3,
            adaptive: Optional[bool] = False,
            rtol: Optional[float] = 1e-5,
            atol: Optional[float] = 1e-4,
            dt_min: Optional[Scalar] = 1e-5,
            options: Optional[Dict[str, Any]] = None,
-           names: Optional[Dict[str, str]] = None) -> TensorOrTensors:
+           names: Optional[Dict[str, str]] = None,
+           **unused_kwargs) -> TensorOrTensors:
```
Comment (patrick-kidger) [inline torchsde/_core/sdeint.py:40]: Returns just a tensor now.

File: torchsde/_core/sdeint.py [hunk 12]
```diff
@@ -93,134 +89,108 @@ def sdeint(sde: [base_sde.BaseSDE],
...
-    return results
 
 
-def check_contract(sde, y0, ts, bm, logqp, method, names):
+def check_contract(sde, y0, ts, bm, method, names):
     if names is None:
         names_to_change = {}
     else:
-        names_to_change = {key: names[key] for key in ('drift', 'diffusion', 'prior_drift') if key in names}
+        names_to_change = {key: names[key] for key in ("drift", "diffusion", "prior_drift") if key in names}
```
Comment (patrick-kidger) [inline torchsde/_core/sdeint.py:99]: No more prior drift.

**Comment by patrick-kidger** [comment]
Also, ` tests/__init__.pyc` shouldn't be included in the commit.

**Comment by patrick-kidger** [comment]
Overall, things are looking so much cleaner now. These changes make me very happy.

File: tests/basic_sde.py [hunk 13]
```diff
@@ -129,11 +129,11 @@ def h(self, t, y):
...
-        self.g_param = nn.Parameter(torch.sigmoid(torch.randn(1, d)), requires_grad=True)
+        self.g_param = nn.Parameter(torch.sigmoid(torch.randn(1, d, 1)), requires_grad=True)
         self.noise_type = "scalar"
 
     def g(self, t, y):
-        return self.g_param.repeat(y.size(0), 1)
+        return self.g_param.repeat(y.size(0), 1, 1)
 
 
 class TupleSDE(SDEIto):
```
[the remark this answers is not in the document]
Comment (lxuechen) [reply tests/basic_sde.py:139]: Done.

File: tests/test_strat.py (hunk 1 above)
Comment (lxuechen) [reply tests/test_strat.py:79]: No, this is simply a rename, i.e. `gdg_jvp` -> `dg_ga_jvp`. I think the latter is clearer, since it matches the math better.

The rationale of including this function is for testing purposes, i.e. we have more confidence in the two variants of jvp I implemented, and it would come in handy in the future if we ever decide to modify any aspect of those functions.

File: torchsde/_core/adjoint.py (hunk 2 above)
Comment (lxuechen) [reply torchsde/_core/adjoint.py:73]: Sure. I'll be using lists for the second part now, as list comprehension is slightly faster.

File: torchsde/_core/base_sde.py (hunk 4 above)
Comment (lxuechen) [reply torchsde/_core/base_sde.py:48]: Done.

File: tests/test_strat.py (hunk 1 above)
Comment (patrick-kidger) [reply tests/test_strat.py:79]: I mean the comment, not the name!

File: torchsde/_core/adjoint.py (hunk 2 above)
Comment (patrick-kidger) [reply torchsde/_core/adjoint.py:73]: Interesting, I didn't know listcomps were faster. Ty.

File: torchsde/_core/sdeint.py (hunk 11 above)
Comment (lxuechen) [reply torchsde/_core/sdeint.py:40]: Done.

File: torchsde/_core/sdeint.py (hunk 10 above)
Comment (lxuechen) [reply torchsde/_core/sdeint.py:29]: Done.

File: torchsde/_core/misc.py (hunk 9 above)
Comment (lxuechen) [reply torchsde/_core/misc.py:140]: Done.

File: torchsde/_core/methods/srk.py (hunk 8 above)
Comment (lxuechen) [reply torchsde/_core/methods/srk.py:72]: I think we should just write wrappers at some point collapsing/uncollapsing the extra dimensions, since writing a single wrapper class is less work than changing all the solvers.

File: torchsde/_core/methods/srk.py (hunk 7 above)
Comment (lxuechen) [reply torchsde/_core/methods/srk.py:63]: Done.

File: torchsde/_core/methods/milstein.py (hunk 6 above)
Comment (lxuechen) [reply torchsde/_core/methods/milstein.py:66]: Done.

File: torchsde/_core/adjoint.py (hunk 3 above)
Comment (lxuechen) [reply torchsde/_core/adjoint.py]: I can change to out-of-place update for now. Though, I'm not inclined to switch from `sdeint.integrate` to `_SdeintAdjointMethod.apply` for now. 

I like testing :), and I think we should do that in another PR after testing that nothing breaks just to be safe.

File: tests/test_strat.py (hunk 1 above)
Comment (lxuechen) [reply tests/test_strat.py:79]: Ah you right, good point!

File: torchsde/_core/adjoint.py (hunk 3 above)
Comment (patrick-kidger) [reply torchsde/_core/adjoint.py]: Fair enough.

File: torchsde/_core/base_sde.py (hunk 5 above)
Comment (patrick-kidger) [reply torchsde/_core/base_sde.py:161]: There's still an unused prior_drift argument.

File: torchsde/_core/methods/srk.py (hunk 8 above)
Comment (patrick-kidger) [reply torchsde/_core/methods/srk.py:74]: Makes sense. We already try to factor as much of the noise type abstraction into the SDE (and out of the solvers) as it is, after all.

**Comment by patrick-kidger** [comment]
Just spotted, both `sdeint` and its adjoint have the docstring for `names` still specifying the possibility of prior drift.

File: torchsde/_core/base_sde.py (hunk 5 above)
Comment (lxuechen) [reply torchsde/_core/base_sde.py:161]: Done.

**Comment by lxuechen** [comment]
> Just spotted, both `sdeint` and its adjoint have the docstring for `names` still specifying the possibility of prior drift.

Done.

**Comment by patrick-kidger** [comment]
This PR looks good to me. I've not hit the merge button as I assume the plan is to merge #40 first, so that the diagnostics can be run.

**Comment by lxuechen** [comment]
Thanks for the reviews @patrick-kidger! All rates preserved after running the diagnostics. Merging now.
